emacs_lua_complete.lua

```diff
@@ -0,0 +1,81 @@
+--    __emacs_lua_complete:
+-- This function finds and prints a limited number of matches on the
+-- lua globals tree, _G. See `lua-complete-string`. N.B.: Does not
+-- complete numeric keys (e.g. for arrays) or keys which begin with an
+-- underscore.
+-- Arguments:
+--   PARTS: a table array of dot-separated parts of the completion item
+--     (e.g. table.con would become {"table","con"}).  All but the last
+--     element must be valid subkeys in the _G heirarchy.
+--   LIBS: an array of tables of the form
+--       {var: varname, lib: libfile}
+--     for e.g. local varname=require("libfile"). This library will be
+--     require'd and searched alongside global variables.
+--   LOCALS: an array of strings parsed from local aloc,bloc = ... to
+--     complete as well
+--
+-- Example:
+-- __emacs_lua_complete({'table','ins'},{{var="xyz",lib="libfile"}},
+--                      {"frank","beans"},5000)
+--
+-- Written 2019/05 J.D. Smith
+function __emacs_lua_complete(parts,libs,locals,limit)
+   -- Setup a FIFO queue for breadth-first tree traversal
+   local queue={head=0,tail=0}
+   function queue:push(value)
+      self.tail=self.tail+1  self[self.tail]=value
+   end
+   function queue:pop()
+      self.head=self.head+1  return self[self.head]
+   end
+
+   -- Mirror the globals table
+   local globals,cnt,limit={},0,limit or 2000
+   for k,v in pairs(_G) do globals[k] = v end
+   
+   -- Add libs and locals
+   if libs then
+      for _,v in ipairs(libs) do
+	 local good,lib=pcall(require,v.lib)
+	 globals[v.var]=good and lib or {}
+      end
+   end
+   if locals then
+      for _,v in ipairs(locals) do globals[v]={} end
+   end
+
+   -- Descend the tree to the limit of dotted parts
+   local g,pre=globals,""
+   for i=1,#parts do
```

Comment:
Would `ipairs` work here to replace all the `parts[i]` in the body of the loop?

Reply:
Well, I use `i` to tell me if I'm in the "last" part, which may be a partial match (or even an empty string for `table.[M-Tab]` type completions).  But I'm by no means a lua expert, so let me know if there's a better lua way.

Reply:
No, what I'm saying is that you are basically doing
```
for i = 1, #parts do
    current_part = parts[i]
```
whereas `ipairs` will do that for you automatically
```
for i, current_part in ipairs(parts)
```

Reply:
Good point, fixed.